File: cockpit/etl/shared/etl_logger.py
from datetime import datetime, timezone
from .supabase_client import get_client

class ETLLogger:
    def __init__(self, source_name: str):
        self.source_name = source_name
        self.started_at = datetime.now(timezone.utc)
        self.log_id: int | None = None

    def start(self):
        client = get_client()
        result = client.table('etl_sync_log').insert({
            'source_name': self.source_name,
            'started_at': self.started_at.isoformat(),
            'status': 'running',
        }).execute()
        if result.data:
            self.log_id = result.data[0]['id']

    def complete(self, rows_upserted: int):
        if not self.log_id:
            return
        now = datetime.now(timezone.utc)
        duration = (now - self.started_at).total_seconds()
        client = get_client()
        client.table('etl_sync_log').update({
            'completed_at': now.isoformat(),
            'status': 'success',
            'rows_upserted': rows_upserted,
            'duration_sec': round(duration, 2),
        }).eq('id', self.log_id).execute()

    def fail(self, error_message: str):
        if not self.log_id:
            return
        now = datetime.now(timezone.utc)
        duration = (now - self.started_at).total_seconds()
        client = get_client()
        client.table('etl_sync_log').update({
            'completed_at': now.isoformat(),
            'status': 'failed',
            'error_message': error_message[:500],
            'duration_sec': round(duration, 2),
        }).eq('id', self.log_id).execute()

File: Tech/CEO-Cockpit/lib/etl/etl-logger.ts
function sbUrl(table: string): string {
  const base = process.env.SUPABASE_URL ?? process.env.NEXT_PUBLIC_SUPABASE_URL!;
  return `${base}/rest/v1/${table}`;
}

function sbHeaders(): Record<string, string> {
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY!;
  return {
    apikey:         key,
    Authorization:  `Bearer ${key}`,
    "Content-Type": "application/json",
    Prefer:         "return=representation",
  };
}

export class ETLLogger {
  private startedAt = new Date();
  private logId: number | null = null;

  constructor(private sourceName: string) {}

  async start(): Promise<void> {
    try {
      const resp = await fetch(sbUrl("etl_sync_log"), {
        method:  "POST",
        headers: sbHeaders(),
        body:    JSON.stringify({
          source_name: this.sourceName,
          started_at:  this.startedAt.toISOString(),
          status:      "running",
        }),
      });
      if (resp.ok) {
        const data = await resp.json() as { id: number }[];
        this.logId = data[0]?.id ?? null;
      }
    } catch { /* ignore logging failures */ }
  }

  async complete(rowsUpserted: number): Promise<void> {
    if (!this.logId) return;
    try {
      const now = new Date();
      await fetch(`${sbUrl("etl_sync_log")}?id=eq.${this.logId}`, {
        method:  "PATCH",
        headers: sbHeaders(),
        body:    JSON.stringify({
          completed_at:  now.toISOString(),
          status:        "success",
          rows_upserted: rowsUpserted,
          duration_sec:  +((now.getTime() - this.startedAt.getTime()) / 1000).toFixed(2),
        }),
      });
    } catch { /* ignore logging failures */ }
  }

  async fail(errorMessage: string): Promise<void> {
    if (!this.logId) return;
    try {
      const now = new Date();
      await fetch(`${sbUrl("etl_sync_log")}?id=eq.${this.logId}`, {
        method:  "PATCH",
        headers: sbHeaders(),
        body:    JSON.stringify({
          completed_at:  now.toISOString(),
          status:        "failed",
          error_message: errorMessage.slice(0, 500),
          duration_sec:  +((now.getTime() - this.startedAt.getTime()) / 1000).toFixed(2),
        }),
      });
    } catch { /* ignore logging failures */ }
  }
}
